x509-examples/signed-02.json

```diff
@@ -0,0 +1,44 @@
+{
+   "title":"signed-02: Signed message w/ x5bag containing two certificates",
+   "input":{
+      "plaintext":"This is the content.",
+      "sign":{
+         "protected":{
+            "ctyp":0
+         },
+         "signers":[
+            {
+               "key":{
+                  "kty":"EC2",
+                  "kid":"Alice Lovelace",
+                  "crv":"P-256",
+                  "x_hex":"863aa7bc0326716aa59db5bf66cc660d0591d51e4891bc2e6a9baff5077d927c",
+		  "y_hex":"ad4eed482a7985be019e9b1936c16e00190e8bcc48ee12d35ff89f0fc7a099ca",
+                  "d_hex":"d42044eb2cd2691e926da4871cf3529ddec6b034f824ba5e050d2c702f97c7a5"
+               },
+               "unprotected":{
+                  "kid":"Alice Lovelace",
+		  "x5bag":[
+                     "308201A930820150A00302010202144E3019548429A2893D04B8EDBA143B8F7D17B276300A06082A8648CE3D040302302C312A30280603550403132153616D706C6520434F534520436572746966696361746520417574686F726974793020170D3230313230323137323732355A180F32303533313031303137323732355A3019311730150603550403130E416C696365204C6F76656C6163653059301306072A8648CE3D020106082A8648CE3D03010703420004863AA7BC0326716AA59DB5BF66CC660D0591D51E4891BC2E6A9BAFF5077D927CAD4EED482A7985BE019E9B1936C16E00190E8BCC48EE12D35FF89F0FC7A099CAA361305F300C0603551D130101FF04023000300F0603551D0F0101FF04050303078000301D0603551D0E041604141151555B01FF3F6DDDF9E5712AD3FF72A2D94D62301F0603551D230418301680141E6FC4D0C0DA004A8427CBBD3FE05A99EA2D2D11300A06082A8648CE3D0403020347003044022038FF9207872BA4D685700774783D35BE5B45AF59265A8567AE952D7182D5CBA00220163A18388EFE6310517385458AB4D3BBF7A0C23D9C87DA1CF378884FBBCDC86C",
+		     "3082019E30820145A003020102021414A4957FD506AA2AAFC669A880032E8C95B87624300A06082A8648CE3D040302302C312A30280603550403132153616D706C6520434F534520436572746966696361746520417574686F726974793020170D3230313230323137323333325A180F32303533313031303137323333325A302C312A30280603550403132153616D706C6520434F534520436572746966696361746520417574686F726974793059301306072A8648CE3D020106082A8648CE3D030107034200047B447C98F731337AFBE3BAC96E793AF12865F3BD56B647A1729764191AE111F3161B4D56FA42F26E1B18DD87F9DB42F4C9168E420E2CE5E2D149648EE0EE5FB4A3433041300F0603551D130101FF040530030101FF300F0603551D0F0101FF04050303070600301D0603551D0E041604141E6FC4D0C0DA004A8427CBBD3FE05A99EA2D2D11300A06082A8648CE3D0403020347003044022006F99B3ACE00007BFB717784DDD230013D8CDCA0BABE20EE00039BEA0898A6D402200FFAF9DE61C1B6BD28BF5DDB1A191E63B22EAD4A69468D5222C487D53C33C204"
+		  ]
```

Comment:
My comment here relates to what I think is an ambiguity in X509 COSE.  I don't think it is clear how the verifier is to determine the end-entity certificate to start the verification with.

It should not be by ordering in x5bag. If it is by order in a list, then it should be x5chain.

It should not be by matching kid with the issuer because nothing says you can or should do that.

And per Ben's comments on x5u and JWS design, it should be by a protected header.

Reply:
I completely agree that ordering in x5bag does not determine which certificate is the end-entity certificate.  I believe that this was by design in order to allow a COSE message to transfer additional certificates (e.g. encryption certificates in a sign message) that the recipient may find useful.  The kid is only for information purposes and RFC8152 says that kid is not a security-critical field and is only a hint about which key to use.  Utilizing such a hint may improve processing, but it must not fail if kid is missing or does not correspond to the certificate SubjectName.  Similarly, an application must not fail to process the signature if the end-entity certificate is not the first certificate in the x5bag.  That does not mean that the first certificate must not be the end-entity certificate or that the kid must not match the SubjectName as in the examples I provided. 

Perhaps you are looking for a more extensive test suite?

Reply:
I agree with everything you say. :-)

But I don't know how to find the end-entity cert in this example.

I think the problem lies in the specifications, not the example.